### DISCUSSION TITLE
Take HTTP methods from files names

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [Danilqa/node-file-router#5] [Ideas]
Hello, congrats on the library!

I played with it a bit, but missed [Nuxt's](https://nuxt.com/docs/guide/directory-structure/server#matching-http-method) way of defining the routes, specifying the HTTP method on the files names like this:

`/api/auth/login.post.ts`.

I couldn't find if that was possible with this library.

So I made my own!

Wrote a function that parses the request, using `node:fs` of course to read recursively the folders and files,
and then parsed the HTTP method from the files names, and it works beautifully.

```ts
export type Context = {
  req: IncomingMessage,
  res: ServerResponse,
  [key: string]: any
}

const httpMethods = [
  'post',
  'get',
  'patch',
  'put',
  'delete',
  'head',
  'options',
] as const

type HttpMethod = 
| Lowercase<typeof httpMethods[number]> 
| Uppercase<typeof httpMethods[number]>


function readFolderRecursively(folderPath: string): string[] {
  const filesArray: string[] = []

  function readFolder(folderPath: string) {
    const files = readdirSync(folderPath)

    files.forEach((file) => {
      const filePath = join(folderPath, file)

      if (statSync(filePath).isDirectory()) {
        readFolder(filePath)
      } else {
        filesArray.push(filePath)
      }
    })
  }

  readFolder(folderPath)
  return filesArray
}


function createRoutes(req: IncomingMessage, res: ServerResponse) {
  let body = ''

  req.on('data', (chunk) => {
    body += chunk
  });

  req.on('end', () => {
    res.setHeader('Content-Type','application/json' )
    
    const files = readFolderRecursively('src/api')
    .map(item => item.replace('src\\api\\', ''))
    .map(item => item.replace('\\', '/'))
    const _res: {
      path?: string
      method: Uppercase<HttpMethod> 
      statusCode: number
      statusMessage: string
      rawFile?: string
    } = {
      path: undefined,
      method: 'GET',
      statusCode: 404,
      statusMessage: 'Route not found',
      rawFile: undefined
    }

    
    for (const rawFile of files) {
      let file: string = rawFile
      let method: Lowercase<HttpMethod>

        
      if (httpMethods.some(item => rawFile.includes(item))) {
        method = httpMethods.find(item => rawFile.includes(item))!
        file = file.replace(`.${method}.ts`, '')

      } else {
        file = file.replace('.ts', '')
        method = 'get'
      }


      if (file === 'index') {
        file = ''
      }

      _res.method = method.toUpperCase() as Uppercase<HttpMethod>
      _res.path = `/${file}`
      _res.rawFile = rawFile

      if (req.url === `/${file}` && method.toUpperCase() === req.method) {
        _res.statusCode = 200
        _res.statusMessage = 'Ok'
        break

      } else if (req.url ===  `/${file}` && method.toUpperCase() !== req.method) {
        _res.statusCode = 404
        _res.statusMessage = `Route ${file} not found`
        break
      }
    }

    res.statusCode = _res.statusCode!
    //res.statusMessage = _res.statusMessage!
    if (_res.statusCode !== 200) {
      res.end(_res.statusMessage)
    } else {
      (async () => {
        const _m = await import(`./api/${_res.rawFile!}`)

        try {
          const result = _m.default({req, res, body})
          if (result.then) {
            return res.end(await result)
          }
          res.end(result)
        } catch (error) {
          if ((error as any).statusCode) {
            res.statusCode = (error as any).statusCode
            res.end((error as any).statusMessage) 
          }
          res.statusCode = 400
          res.end((error as Error).message)
        }
      })()
    } 
  })
}

const server: Server = createServer((req: IncomingMessage, res: ServerResponse) => {
  createRoutes(req, res)
})



server.listen(5000, () => {
  console.log('Listening')
})
```
![image](https://github.com/Danilqa/node-file-router/assets/10191085/e8e99c12-7a89-4827-90fa-4344d0c13b00)

Obviously it's very raw, maybe I polish it a bit later, work on performance, and make the paths work with other operating systems (currently made it for Windows!).
 You can implement this feature if you want, it would be cool.

Wishing you health, peace, and success. 🔥☕😊🤝

**Comment by Danilqa** [comment]
@angelhdzmultimedia 

Hi! 👋🏻

Wow, your idea is brilliant!! 🔥 Thank you so much. It will definitely become a part of the functionality.

I'll do it on this Tuesday or Wednesday :) And notify you when it's ready. 

Also, if you suddenly have the time and desire, I can help you with MR. I’m ready to provide guidance on the code as to what is needed and how it could be roughly implemented. It has several nuances: the library works not only with http but with any other protocols.

**Comment by Danilqa** [comment]
@angelhdzmultimedia Hi! I've released a new version with this functionality. 

https://github.com/Danilqa/node-file-router/releases/tag/v0.3.0

**Reply by angelhdzmultimedia** [reply]
My apologies! I got driven away by the hype, and was trying it on stackblitz, that I forgot to close the discussion!
Gave you more work to do that the one you have done already!
Appreciate you taking my idea into consideration, and even including it lightning fast! 🔥☕😊🎉⚡🏃‍♂️👏

**Reply by Danilqa** [reply]
No problem at all! I noticed the smiles and guessed that everything is alright. Thank you so much again, I'm super happy now! :) If anything, I can open it again 🙂
